src/main/scala/tech/cryptonomic/conseil/generic/rpc/DataFetcher.scala

```diff
@@ -0,0 +1,189 @@
+  * of each step, and an example from tezos
+  *
+  * [Inputs]               -> [Fetch a Data Batch]                                                        -> [Decode Data]
+  * [A list of input data] -> [A list of json data paired with corresponding input, wrapped in an effect] -> [A list of output data, paired with inputs, wrapped in the effect]
+  * [List of offsets]      -> [Future of pairs (offset, json) that describes block data]                  -> [Future of pairs (offset, BlockData) with the block now decoded]
+  * [List of block hashes] -> [Future of pairs (hash, json) that describes operations data]               -> [Future of pairs (hash, List[Operations]) with the ops now decoded]
+  *
+  * which suggest that we may encode this in a trait that separates the steps with a
+  * - fetchData function, that, given a collection of inputs, returns the effectful result
+      of encoded data (e.g. json), paired with its input value (to preserve correlation)
+  * - decodeData function that will decode every single endoded value to a full-blown output scala data type
+  *
+  * The effectul function is defined in term of a "Kleisli arrow", which is nothing but a fancy-named wrapper for A => F[B], where F[_] is the so-called effect
+  * Using Kleisli enables a plethora of useful methods to call, that will more easily convert a Kleisli arrow to something that accepts different input, or that
+  * adds further operations on input or output, or that adds other effects, like traversing or folding a collection of results if the B result type allows that.
+  *
+  * This enables us to easily compose the `fetchData` function and `decodeData` function to a general `fetch` function, that will use the former to retrieve a list of
+  * encoded values, paired with input, and apply the decoding to each, all the while preserving the correlation with the original input value.
+  *
+  * In addition to this, we might want to apply multiple decoding steps in parallel, and pair the decoded results with the same inputs
+  * Or still we might want to apply more than one fetcher to the same set of inputs, and pair all the results, in a single pass
+  * This is where the `addDecoding` and `mergeResults`/`fetchCombine` combinators comes useful.
+  */
+object DataFetcher {
+  import cats.syntax.applicativeError._
+  import cats.syntax.semigroupal._
+
+  /** using the combination of the provided fetcher functions we fetch all data with this general outline:
+    * 1. get encoded (e.g. json) values from the node, given a traversable input collection
+    * 2. keep the correlation with the input and decode each value (e.g. from json) to a corresponding output
+    * 3. return the traversable collection containing paired input and output values, all wrapped in the effect F
+    * The resulting effect will also encode a possible failure in virtue of the implicit MonadError instance that is provided
+    */
+  def fetcher[Eff[_], In, Out, Err](
```

Comment:
How come this functions name is a noun rather than a verb?

Reply:
It slipped by with the rush of completing the work.

I had a hard time figuring out a good name for this, I'll change to a getter first thing: `getFetcher`